onnxruntime/core/providers/shared_library/provider_bridge_provider.cc

```diff
@@ -0,0 +1,287 @@
+
+#define PROVIDER_NOT_IMPLEMENTED ORT_THROW("Unimplemented shared library provider method");
+
+namespace onnxruntime {
+
+void SetProviderHost(ProviderHost& host) {
+  g_host = &host;
+}
+
+struct OnUnloadFunction {
+  OnUnloadFunction(std::function<void()> function) : function_(std::move(function)) {}
+
+  std::function<void()> function_;
+  bool enabled_{true};
+};
+
+static std::unique_ptr<std::vector<std::unique_ptr<OnUnloadFunction>>> s_run_on_unload_;
+
+RunOnUnload::RunOnUnload(std::function<void()> deleter) {
+  static std::mutex mutex;
+  std::lock_guard<std::mutex> guard{mutex};
+  if (!s_run_on_unload_)
+    s_run_on_unload_ = onnxruntime::make_unique<std::vector<std::unique_ptr<OnUnloadFunction>>>();
+  auto unload_function = std::make_unique<OnUnloadFunction>(std::move(deleter));
+  enabled_ = &unload_function->enabled_;
+  s_run_on_unload_->push_back(std::move(unload_function));
+}
+
+RunOnUnload::~RunOnUnload() {
+  *enabled_ = false;  // If the thread_local gets destroyed, then disalble the delete function
+}
+
+struct OnUnload {
+  ~OnUnload() {
+    if (!s_run_on_unload_)
+      return;
+
+    for (auto& function : *s_run_on_unload_) {
+      if (function->enabled_)
+        function->function_();
+    }
+
+    s_run_on_unload_.reset();
+  }
+
+} g_on_unload;
+
+}  // namespace onnxruntime
+
+// Override default new/delete so that we match the host's allocator
+void* operator new(size_t n) { return g_host->HeapAllocate(n); }
+void operator delete(void* p) { return g_host->HeapFree(p); }
+void operator delete(void* p, size_t /*size*/) { return g_host->HeapFree(p); }
+
+namespace onnx {
+std::unique_ptr<ONNX_NAMESPACE::Provider_AttributeProto> Provider_AttributeProto::Create() {
+  return g_host->AttributeProto_Create();
+}
+}  // namespace onnx
+
+namespace onnxruntime {
+
+Provider_AllocatorPtr CreateAllocator(Provider_DeviceAllocatorRegistrationInfo& info, int16_t device_id) {
+  return g_host->CreateAllocator(info, device_id);
+}
+
+std::unique_ptr<Provider_KernelDefBuilder> Provider_KernelDefBuilder::Create() {
+  return g_host->KernelDefBuilder_Create();
+}
+
+std::shared_ptr<Provider_KernelRegistry> Provider_KernelRegistry::Create() {
+  return g_host->KernelRegistry_Create();
+}
+
+std::unique_ptr<Provider_OrtMemoryInfo> Provider_OrtMemoryInfo::Create(const char* name_, OrtAllocatorType type_, Provider_OrtDevice* device_, int id_, OrtMemType mem_type_) {
+  return g_host->OrtMemoryInfo_Create(name_, type_, device_, id_, mem_type_);
+}
+
+std::unique_ptr<Provider_IndexedSubGraph> Provider_IndexedSubGraph::Create() {
+  return g_host->IndexedSubGraph_Create();
+}
+
+template <>
+MLDataType DataTypeImpl::GetType<float>() {
+  return g_host->DataTypeImpl_GetType_float();
+}
+
+template <>
+MLDataType DataTypeImpl::GetTensorType<float>() {
+  return g_host->DataTypeImpl_GetTensorType_float();
+}
+
+TensorShape::TensorShape(const int64_t* dimension_sizes, size_t dimension_count)
+    : std::vector<int64_t>(dimension_count) {
```

Comment:
Instead of duplicating code can we put the implementation in a .cc file that gets #included?

Reply:
It might be difficult since the internal code likes to include other things. This would need to have zero dependencies on the other internal headers (besides a couple that are very limited). I don't like this either.